feat(provider/azure): Add Power BI triggers, operators and tests for dataset and workspace operations

mvfc opened the pull request:
## Description
This PR introduces deferrable versions of the Power BI workspace- and dataset-list operators in the Azure provider package. Instead of blocking a worker slot while calling the Power BI REST API, these operators now:

- **Defer** execution using custom `BaseTrigger` implementations  
- **Poll** asynchronously for the full list of workspaces or datasets  
- **Resume** execution once data is available and push IDs into XCom  

### New Hook methods
- `PowerBIHook.get_workspace_list()`  
- `PowerBIHook.get_dataset_list(group_id: str)`

### New Trigger classes
- `PowerBIWorkspaceListTrigger`  
- `PowerBIDatasetListTrigger`

### New Operator classes
- `PowerBIWorkspaceListOperator`  
- `PowerBIDatasetListOperator`

### Unit tests
- Trigger serialization (`serialize()`)  
- Trigger run logic (success and error paths)  
- Operator deferral (`TaskDeferred`)  
- Operator resume handling (`execute_complete`)  
- XCom key/value assertions  

## Motivation and Context
Long-running or paginated API calls can occupy a worker slot for tens of seconds or more, limiting concurrency and throughput. By deferring:

1. We free up the worker while waiting on Power BI.  
2. We align with Airflow’s best practices for asynchronous I/O.  
3. We make scaling in high-volume environments more efficient.

This change complements existing synchronous operators; it does not remove or alter any blocking behavior, giving users the choice.

## How Has This Been Tested?
- 🧪 **Automated tests** under `tests/providers/microsoft/azure/` covering all new code paths.  
- 🔄 **Manual end-to-end DAG** against a real Power BI service principal.  
- ✅ Verified that XCom payloads contain the correct ID lists on resume.  
- ✅ Verified appropriate error handling and retries in failure cases.

## Documentation
- Added `howto/operator:PowerBIWorkspaceListOperator` and `howto/operator:PowerBIDatasetListOperator` pages.  
- Updated Azure provider README to mention the new deferrable operators.

## Backwards-Incompatible Changes?
No. Existing synchronous `PowerBIHook` methods and operators remain unchanged and fully supported.

## New Dependencies?
None. All new code relies on existing Airflow and Azure provider infrastructure.

jscheffl commented on dev/breeze/README.md:
```diff
@@ -135,6 +135,6 @@ PLEASE DO NOT MODIFY THE HASH BELOW! IT IS AUTOMATICALLY UPDATED BY PRE-COMMIT.
 ---------------------------------------------------------------------------------------------------------
 
-Package config hash: bf3faa4e35a6ee2b2bc45ab87d79b5050d443ebffc00ea95402af12ab5345252ae3e5abb489b6d2f65ef146de4fb66e1c145c6ad48fc27b916782c163cb5db20
+Package config hash: 0cfb22ca168edc3bdf44937124549fa798a97f0641429d565e50bffa1ae3ed3142e9459f8b2ca7b1ddab8797f5527e5c19d9ecdf1f11231c533a3fe7d865b1e2
```
Why actually in your PR is the config hash updated? Does not seem that you touch anything in breeze.

mvfc replied:
I don’t know, I ran pre-commit and it updated it. Thought that may have been the intended consequence. Let me revert that part.

jscheffl replied:
Maybe you need to update your local breeze and an outdated version creates a "wrong" hash?

mvfc replied:
Possibly. Tomorrow morning I will try to upgrade breeze, thanks for the tip. I reverted that one.

mvfc committed:
feat(powerbi): add deferrable PowerBI workspace & dataset list operators and triggers

- Introduce PowerBIWorkspaceListOperator and PowerBIDatasetListOperator that defer using triggers
- Implement PowerBIWorkspaceListTrigger and PowerBIDatasetListTrigger for async polling of list endpoints
- Hook into the Microsoft Graph API via new async methods on PowerBIHook (get_workspace_list, get_dataset_list)
- Add XCom support in execute_complete callbacks for both operators
- Include unit tests covering deferral, trigger serialization, run logic, and operator resume paths
- Update docs: operator HOWTO for PowerBI workspace/dataset listing

gopidesupavan commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/operators/powerbi.py:
```diff
@@ -167,3 +171,182 @@ def execute_complete(self, context: Context, event: dict[str, str]) -> Any:
+        self,
+        *,
+        conn_id: str = PowerBIHook.default_conn_name,
+        timeout: float = 60 * 60 * 24 * 7,
```
I would suggest here keep complete value instead of evaluating here and may be add a docstring how that value added.

gopidesupavan replied:
seems docstrings are missing for other params, can you please add them.

mvfc replied:
> I would suggest here keep complete value instead of evaluating here and may be add a docstring how that value added.

but what if someone changes the default connection name for some reason? it wont be in sync then, wouldn't it be better to evaluate here rather than hardcode it again?

gopidesupavan replied:
my bad not for `conn_id` i am referring. am referring to `timeout` value

mvfc replied:
@gopidesupavan Oh the timeout value I just took what @ambika-garg used on the other operators. Took it to be the Microsoft default for None (a week). Just left it there to be inline with the other existing operator, triggers and hook.

gopidesupavan commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/operators/powerbi.py:
```diff
@@ -167,3 +171,182 @@ def execute_complete(self, context: Context, event: dict[str, str]) -> Any:
+            method_name=self.execute_complete.__name__,
+        )
+
+    def get_workspace_list(
```
Not sure where is this method being used?

mvfc replied:
removed, was supposed to be for testing, same as the other one

gopidesupavan commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/operators/powerbi.py:
```diff
@@ -167,3 +171,182 @@ def execute_complete(self, context: Context, event: dict[str, str]) -> Any:
+            method_name=self.execute_complete.__name__,
+        )
+
+    def get_dataset_list(self, context: Context, event: dict[str, str] | dict[str, list[str]] | None = None):
```
Same here aswell, where this being used?

mvfc replied:
removed, was supposed to be only for testing

gopidesupavan commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/operators/powerbi.py:
```diff
@@ -167,3 +171,182 @@ def execute_complete(self, context: Context, event: dict[str, str]) -> Any:
+        group_id: str,
+        conn_id: str = PowerBIHook.default_conn_name,
+        timeout: float = 60 * 60 * 24 * 7,
+        proxies: dict | None = None,
```
Please update docstrings.

mvfc replied:
updated

gopidesupavan commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/powerbi.py:
```diff
@@ -230,3 +230,165 @@ async def fetch_refresh_status_and_error() -> tuple[str, str]:
+                    "workspace_ids": None,
+                }
+            )
+            return
```
When condition not met here this will not yield anything,  Triggers should yield with `TriggerEvent`, please add some information and return.

gopidesupavan commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/powerbi.py:
```diff
@@ -230,3 +230,165 @@ async def fetch_refresh_status_and_error() -> tuple[str, str]:
+                    "dataset_ids": None,
+                }
+            )
+            return
```
Same as above comment, it should yield something if not met any condition.

mvfc replied:
conditions weren't supposed to be there anymore, it was for another test i was doing locally, my bad. thanks for catching this.

mvfc committed:
removing unnecessary conditional tests on triggers

gopidesupavan reviewed (approved):
Good work, thanks for the changes.
